File: WeatherStation/CS/WeatherStation/StartupTask.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net.Http;
using Windows.ApplicationModel.Background;
using Windows.Devices.I2c;
using Windows.Devices.Enumeration;
using Windows.System.Threading;


namespace WeatherStation
{
    public sealed class StartupTask : IBackgroundTask
    {
        BackgroundTaskDeferral deferral;
        private ThreadPoolTimer timer;
        I2cDevice sensor;

        public async void Run(IBackgroundTaskInstance taskInstance)
        {
            deferral = taskInstance.GetDeferral();

            String aqs = I2cDevice.GetDeviceSelector("I2C1");
            IReadOnlyList<DeviceInformation> dis = await DeviceInformation.FindAllAsync(aqs);
            //Ox40 was determined by looking at the datasheet for the device
            sensor = await I2cDevice.FromIdAsync(dis[0].Id, new I2cConnectionSettings(0x40));
               
            timer = ThreadPoolTimer.CreatePeriodicTimer(Timer_Tick, TimeSpan.FromMilliseconds(500));

        }

        private void Timer_Tick(ThreadPoolTimer timer)
        {
            //Read temperature and humidity and send result to debugger output window

            byte[] tempCommand = new byte[1] { 0xE3 };
            byte[] tempData = new byte[2];
            sensor.WriteRead(tempCommand, tempData);
            var rawTempReading = tempData[0] << 8 | tempData[1];
            var tempRatio = rawTempReading / (float)65536;
            double temperature = (-46.85 + (175.72 * tempRatio)) * 9 / 5 + 32;
            System.Diagnostics.Debug.WriteLine("Temp: " + temperature.ToString());

            byte[] humidityCommand = new byte[1] { 0xE5 };
            byte[] humidityData = new byte[2];
            sensor.WriteRead(humidityCommand, humidityData);
            var rawHumidityReading = humidityData[0] << 8 | humidityData[1];
            var humidityRatio = rawHumidityReading / (float)65536;
            double humidity = -6 + (125 * humidityRatio);
            System.Diagnostics.Debug.WriteLine("Humidity: " + humidity.ToString());

        }
    }
}

File: WeatherStation/Cpp/WeatherStation/StartupTask.cpp
// Copyright (c) Microsoft. All rights reserved.

#include "pch.h"
#include "StartupTask.h"

using namespace WeatherStation;

using namespace Platform;
using namespace Windows::ApplicationModel::Background;
using namespace Windows::Foundation;
using namespace Windows::Devices::I2c;
using namespace Windows::Devices::Enumeration;
using namespace Windows::System::Threading;
using namespace concurrency;

StartupTask::StartupTask()
{
}

void StartupTask::Run(IBackgroundTaskInstance^ taskInstance)
{
	Deferral = taskInstance->GetDeferral();

	auto controllerTask = create_task(I2cController::GetDefaultAsync());
	controllerTask.then([this](I2cController^ controller) {
		Device = controller->GetDevice(ref new I2cConnectionSettings(0x40));
	});
	


	TimerElapsedHandler ^handler = ref new TimerElapsedHandler(
		[this](ThreadPoolTimer ^timer)
	{
		wchar_t humidityOutput[100];
		wchar_t temperatureOutput[100];

		double humidity = GetHumidity();
		swprintf_s(humidityOutput, 100, L"Humidity: %f\n", humidity);
		OutputDebugStringW(humidityOutput);

		
		double temperature = GetTemperature();
		swprintf_s(temperatureOutput, 100, L"Temperature: %f\n", temperature);
		OutputDebugStringW(temperatureOutput);



		
	});

	TimeSpan interval;
	interval.Duration = 1000 * 1000 * 10;
	Timer = ThreadPoolTimer::CreatePeriodicTimer(handler, interval);
}

double StartupTask::GetTemperature() 
{
	Platform::Array<byte>^ command = ref new Platform::Array<byte>(1);
	command[0] = 0xE3;
	Array<byte>^ data = ref new Array<byte>(2);
	Device->WriteRead(command, data);
	auto rawReading = data[0] << 8 | data[1];
	auto ratio = rawReading / (float)65536;
	double temperature = (-46.85 + (175.72 * ratio)) * 9 / 5 + 32;
	return temperature;
}

double StartupTask::GetHumidity()
{
	Platform::Array<byte>^ command = ref new Platform::Array<byte>(1);
	command[0] = 0xE5;
	Array<byte>^ data = ref new Array<byte>(2);
	Device->WriteRead(command, data);
	auto rawReading = data[0] << 8 | data[1];
	auto ratio = rawReading / (float)65536;
	double humidity = -6 + (125 * ratio);
	return humidity;
}